cpp/include/patterns/copy_v_transform_reduce_nbr.cuh

```diff
@@ -0,0 +1,293 @@
+ * and thrust::copy() (update vertex properties part, take transform_reduce output as copy input).
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @tparam AdjMatrixRowValueInputIterator Type of the iterator for graph adjacency matrix row
+ * input properties.
+ * @tparam AdjMatrixColValueInputIterator Type of the iterator for graph adjacency matrix column
+ * input properties.
+ * @tparam EdgeOp Type of the binary (or ternary) edge operator.
+ * @tparam T Type of the initial value for reduction over the incoming edges.
+ * @tparam VertexValueOutputIterator Type of the iterator for vertex output property variables.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param adj_matrix_row_value_input_first Iterator pointing to the adjacency matrix row input
+ * properties for the first (inclusive) row (assigned to this process in multi-GPU).
+ * `adj_matrix_row_value_input_last` (exclusive) is deduced as @p adj_matrix_row_value_input_first +
+ * @p graph_view.get_number_of_adj_matrix_local_rows().
+ * @param adj_matrix_col_value_input_first Iterator pointing to the adjacency matrix column input
+ * properties for the first (inclusive) column (assigned to this process in multi-GPU).
+ * `adj_matrix_col_value_output_last` (exclusive) is deduced as @p adj_matrix_col_value_output_first
+ * + @p graph_view.get_number_of_adj_matrix_local_cols().
+ * @param e_op Binary (or ternary) operator takes *(@p adj_matrix_row_value_input_first + i), *(@p
+ * adj_matrix_col_value_input_first + j), (and optionally edge weight) (where i and j are row and
+ * column indices, respectively) and returns a value to be reduced.
+ * @param init Initial value to be added to the reduced @e_op return values for each vertex.
+ * @param vertex_value_output_first Iterator pointing to the vertex property variables for the first
+ * (inclusive) vertex (assigned to tihs process in multi-GPU). `vertex_value_output_last`
+ * (exclusive) is deduced as @p vertex_value_output_first + @p
+ * graph_view.get_number_of_local_vertices().
+ */
+template <typename GraphViewType,
+          typename AdjMatrixRowValueInputIterator,
+          typename AdjMatrixColValueInputIterator,
+          typename EdgeOp,
+          typename T,
+          typename VertexValueOutputIterator>
+void copy_v_transform_reduce_in_nbr(raft::handle_t const& handle,
+                                    GraphViewType const& graph_view,
+                                    AdjMatrixRowValueInputIterator adj_matrix_row_value_input_first,
+                                    AdjMatrixColValueInputIterator adj_matrix_col_value_input_first,
+                                    EdgeOp e_op,
+                                    T init,
+                                    VertexValueOutputIterator vertex_value_output_first)
```

Comment:
Suggest moving the output iterator to be the fifth parameter (before `EdgeOp e_op`).  This would be more consistent with `thrust::transform` and `thrust::copy_if`.

Reply:
So, this has two parts, copy_v and transform_reduce_in_nbr.

```raft::handle_t const& handle``` and ```GraphViewType const& graph_view``` are for both.

```AdjMatrixRowValueInputIterator adj_matrix_row_value_input_first```, ```AdjMatrixColValueInputIterator adj_matrix_col_value_input_first```, ```EdgeOp e_op```, and ```T init``` are for transform_reduce_in_nbr.

```VertexValueOutputIterator vertex_value_output_first``` is for copy_v (copying the output of transform_reduce_in_nbr for each vertex to the memory pointed by this output iterator.

So, moving ```vertex_value_output_first``` before ```e_op``` places a parameter for ```copy_v``` in the middle of parameters for ```transform_reduce_i_nbr```. This being said, do you still recommend moving?

Reply:
If I compare this to thrust::copy, 

```AdjMatrixRowValueInputIterator adj_matrix_row_value_input_first```, ```AdjMatrixColValueInputIterator adj_matrix_col_value_input_first```, ```EdgeOp e_op```, and ```T init``` are ```InputIterator first``` and ```InputIterator last``` gets deduced from ```first``` and ```graph_view```.

Reply:
I think of this functionality as a transform rather than a copy, where the transform operation is a reduction over the neighbors of each vertex.  So my intuition was to think of this as a fancy thrust::transform rather than a copy.  The transform API specifies the input and output iterators first and the transformation details last.

Of course, your naming of the function suggests copy of a reduction, which is obviously how you were thinking of the function.

I think it's fine either way.  It just was not quite what I expected.  I suppose if you were to change to specifying them like a transform it might be better to name the function differently (with the transform first to emphasis that intended focus).

Reply:
OK, make sense, and I will keep this as is for now.